src/ocrd/processor/base.py

```diff
@@ -188,20 +233,137 @@ def verify(self):
         """
         return True
 
+    def setup(self) -> None:
+        """
+        Prepare the processor for actual data processing,
+        prior to changing to the workspace directory but
+        after parsing parameters.
+
+        (Override this to load models into memory etc.)
+        """
+        pass
+
+    @deprecated(version='3.0', reason='process() should be replaced with process_page() and process_workspace()')
     def process(self) -> None:
         """
-        Process the :py:attr:`workspace` 
+        Process all files of the :py:attr:`workspace` 
         from the given :py:attr:`input_file_grp`
         to the given :py:attr:`output_file_grp`
-        for the given :py:attr:`page_id`
+        for the given :py:attr:`page_id` (or all pages)
         under the given :py:attr:`parameter`.
         
         (This contains the main functionality and needs to be overridden by subclasses.)
         """
         raise NotImplementedError()
 
+    def process_workspace(self, workspace: Workspace) -> None:
+        """
+        Process all files of the given ``workspace``,
+        from the given :py:attr:`input_file_grp`
+        to the given :py:attr:`output_file_grp`
+        for the given :py:attr:`page_id` (or all pages)
+        under the given :py:attr:`parameter`.
+
+        (This will iterate over pages and files, calling
+        :py:meth:`process_page`, handling exceptions.)
+        """
+        # assert self.input_file_grp is not None
+        # assert self.output_file_grp is not None
+        # input_file_grps = self.input_file_grp.split(',')
+        # for input_file_grp in input_file_grps:
+        #     assert input_file_grp in workspace.mets.file_groups
+        log = getLogger('ocrd.processor.base')
+        with pushd_popd(workspace.directory):
+            self.workspace = workspace
+            try:
+                # FIXME: add page parallelization by running multiprocessing.Pool (#322)
+                for input_file_tuple in self.zip_input_files(on_error='abort'):
```

Comment:
No, none ATM. That choice only about the edge case of getting multiple OcrdFile results per page (aside the single-PAGE multiple images logic).

Reply:
Oops! Not true anymore. That narrow meaning has [recently been conflated](https://github.com/OCR-D/core/commit/64b356384be8155797ae6147f23447773992c2b0) with the behaviour in case no files match for the specified pageId. It also conflates `last=first=skip`.

I don't remember where we wanted to situate error handling for malformed pageId (ranges), but it feels wrong to conflate these cases. One is more of a METS validity problem, the other more of a suboptimal runtime choice.

> Because that might be a better default for any processors that still rely on `process` and iterating themselves.

I don't quite understand that part, sry.

Reply:
>> Because that might be a better default for any processors that still rely on `process` and iterating themselves. 
> I don't quite understand that part, sry.

I mean that there are processors that use `zip_input_files()`, ie. without overriding `on_error`. E.g. in `ocrd-segment-replace-page`, which will silently skip over results if there's no two matching files for a page. Which might be the desired behavior in this case but I was wondering whether the default behavior should be to be strict and if processor developer wants the "silently skip" behavior, `on_error` should be overridden explicitly instead of the other way around.